wrapper/src/main/java/software/amazon/jdbc/util/SlidingExpirationCache.java

```diff
@@ -118,10 +118,15 @@ protected void removeAndDispose(K key) {
   }
 
   protected void removeIfExpired(K key) {
-    final CacheItem cacheItem = cache.get(key);
-    if (cacheItem == null || cacheItem.shouldCleanup()) {
-      removeAndDispose(key);
-    }
+    // TODO: what if the disposal function is long-running? It will lock up the cache until its done.
+    cache.computeIfPresent(key, (k, cacheItem) -> {
+      if (cacheItem.shouldCleanup()) {
+        removeAndDispose(key);
+        return null;
+      }
+
+      return cacheItem;
+    });
```

Comment:
This change was introduced because the shouldCleanup check and the removal of the item from the cache need to be synchronized. However, my concern with this change is that some disposal functions may run for a long time and the cache will be locked during this time, for example the failover2 plugin attempts to clean up its threads for up to 30 seconds. I could implement an additional change outlined below but I'm not sure if there's a better way. The additional change moves the disposal call until after the item has been removed from the cache and the cache has been unlocked.
```suggestion
    final AtomicReference<V> itemRef = new AtomicReference<>();
    cache.computeIfPresent(key, (k, cacheItem) -> {
      if (cacheItem.shouldCleanup()) {
        cache.remove(key);
        itemRef.set(cacheItem.item);
        return null;
      }

      return cacheItem;
    });

    V item = itemRef.get();
    if (item != null && itemDisposalFunc != null) {
      itemDisposalFunc.dispose(item);
    }
```

Reply:
That's a valid concern. I'd just simplify it a bit. I don't think we need weak pointers here. A regular list (ArrayList) would be just fine. Line 125 (cache.remove) is unnecessary since your lambda return null that equals to item deletion.

Reply:
This fix was merged into main in another PR, ended up using an ArrayList